[CI] GeoHexTilerTests testGeoGridSetValuesBoundingBoxes_UnboundedGeoShapeCellValues failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/n5u3dbgtyfcya/tests/:x-pack:plugin:spatial:test/org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoHexTilerTests/testGeoGridSetValuesBoundingBoxes_UnboundedGeoShapeCellValues

**Reproduction line:**
```
./gradlew ':x-pack:plugin:spatial:test' --tests "org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoHexTilerTests.testGeoGridSetValuesBoundingBoxes_UnboundedGeoShapeCellValues" -Dtests.seed=1979F5B25606B4A9 -Dtests.locale=uk-UA -Dtests.timezone=US/Samoa -Druntime.java=17
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoHexTilerTests&tests.test=testGeoGridSetValuesBoundingBoxes_UnboundedGeoShapeCellValues

**Failure excerpt:**
```
java.lang.AssertionError: [75:0] bucket count
Expected: <5>
     but: was <4>

  at __randomizedtesting.SeedInfo.seed([1979F5B25606B4A9:6C4FC5FF63327B8F]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoGridTilerTestCase.testGeoGridSetValuesBoundingBoxes_UnboundedGeoShapeCellValues(GeoGridTilerTestCase.java:176)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:833)

```

--- a/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java
+++ b/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java
@@ -8,6 +8,7 @@
 package org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid;
 
 import org.elasticsearch.h3.H3;
+import org.elasticsearch.xpack.spatial.common.H3CartesianUtil;
 import org.elasticsearch.xpack.spatial.index.fielddata.GeoRelation;
 import org.elasticsearch.xpack.spatial.index.fielddata.GeoShapeValues;
 
@@ -19,15 +20,6 @@ import java.io.IOException;
 abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
 
     private static final long[] RES0CELLS = H3.getLongRes0Cells();
-    // The hexRing neighbours optimization is insufficient at the Poles, so we do a little extra checking
-    private static final long[] NORTH_POLAR_CELLS = new long[H3.MAX_H3_RES + 1];
-    private static final long[] SOUTH_POLAR_CELLS = new long[H3.MAX_H3_RES + 1];
-    static {
-        for (int res = 0; res <= H3.MAX_H3_RES; res++) {
-            NORTH_POLAR_CELLS[res] = H3.geoToH3(90, 0, res);
-            SOUTH_POLAR_CELLS[res] = H3.geoToH3(-90, 0, res);
-        }
-    }
 
     AbstractGeoHexGridTiler(int precision) {
         super(precision);
@@ -100,7 +92,7 @@ abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
             // Normally sufficient to check only bottom-left against top-right
             return -1;
         }
-        if (minH3 == NORTH_POLAR_CELLS[res] || minH3 == SOUTH_POLAR_CELLS[res]) {
+        if (H3CartesianUtil.isPolar(minH3)) {
             // But with polar cells we must check the other two corners too
             final long minMax = H3.geoToH3(bounds.minY(), bounds.maxX(), res);
             final long maxMin = H3.geoToH3(bounds.maxY(), bounds.minX(), res);

--- a/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java
+++ b/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java
@@ -19,6 +19,11 @@ import java.io.IOException;
 abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
 
     private static final long[] RES0CELLS = H3.getLongRes0Cells();
+    private static final long SOUTH_POLE = H3.geoToH3(-90, 0, 0);
+    private static final long NORTH_POLE = H3.geoToH3(90, 0, 0);
+    // The hexRing neighbours optimization is insufficient at the Poles, so we add in one more neighbour explicitly
+    private static final long SOUTH_POLE_EXTRA_NEIGHBOUR = 580542139465727999L;
+    private static final long NORTH_POLE_EXTRA_NEIGHBOUR = 576707042908045311L;
 
     AbstractGeoHexGridTiler(int precision) {
         super(precision);
@@ -74,7 +79,7 @@ abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
         // Point resolution is done using H3 library which uses spherical geometry. It might happen that in cartesian, the
         // actual point value is in a neighbour cell as well.
         {
-            for (long n : H3.hexRing(h3)) {
+            for (long n : hexRing(h3)) {
                 final GeoRelation relation = relateTile(geoValue, n);
                 valueIndex = maybeAdd(n, relation, values, valueIndex);
                 if (relation == GeoRelation.QUERY_CONTAINS) {
@@ -83,6 +88,26 @@ abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
             }
         }
         return valueIndex;
+    }
+
+    private static long[] hexRing(long h3) {
+        long[] neighbours = H3.hexRing(h3);
+        if (h3 == NORTH_POLE) {
+            // Due to equi-rectangular distortion, the hexRing neighbours of the South Pole are insufficient
+            return addNeighbour(neighbours, NORTH_POLE_EXTRA_NEIGHBOUR);
+        } else if (h3 == SOUTH_POLE) {
+            // Due to equi-rectangular distortion, the hexRing neighbours of the South Pole are insufficient
+            return addNeighbour(neighbours, SOUTH_POLE_EXTRA_NEIGHBOUR);
+        } else {
+            return neighbours;
+        }
+    }
+
+    private static long[] addNeighbour(long[] neighbours, long extraNeighbour) {
+        long[] extended = new long[neighbours.length + 1];
+        System.arraycopy(neighbours, 0, extended, 0, neighbours.length);
+        extended[neighbours.length] = extraNeighbour;
+        return extended;
     }
 
     /**
@@ -110,8 +135,9 @@ abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
             final long minH3 = H3.geoToH3(bounds.minY(), bounds.minX(), 0);
             final long maxH3 = H3.geoToH3(bounds.maxY(), bounds.maxX(), 0);
             if (minH3 == maxH3) {
+                // When the level 0 bounds are within a single cell, we can search that cell and its immediate neighbours
                 valueIndex = setValuesByRecursion(values, geoValue, minH3, 0, valueIndex);
-                for (long n : H3.hexRing(minH3)) {
+                for (long n : hexRing(minH3)) {
                     valueIndex = setValuesByRecursion(values, geoValue, n, 0, valueIndex);
                 }
                 return valueIndex;

--- a/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java
+++ b/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java
@@ -19,11 +19,15 @@ import java.io.IOException;
 abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
 
     private static final long[] RES0CELLS = H3.getLongRes0Cells();
-    private static final long SOUTH_POLE = H3.geoToH3(-90, 0, 0);
-    private static final long NORTH_POLE = H3.geoToH3(90, 0, 0);
-    // The hexRing neighbours optimization is insufficient at the Poles, so we add in one more neighbour explicitly
-    private static final long SOUTH_POLE_EXTRA_NEIGHBOUR = 580542139465727999L;
-    private static final long NORTH_POLE_EXTRA_NEIGHBOUR = 576707042908045311L;
+    // The hexRing neighbours optimization is insufficient at the Poles, so we do a little extra checking
+    private static final long[] NORTH_POLAR_CELLS = new long[H3.MAX_H3_RES];
+    private static final long[] SOUTH_POLAR_CELLS = new long[H3.MAX_H3_RES];
+    static {
+        for (int res = 0; res < H3.MAX_H3_RES; res++) {
+            NORTH_POLAR_CELLS[res] = H3.geoToH3(90, 0, res);
+            SOUTH_POLAR_CELLS[res] = H3.geoToH3(-90, 0, res);
+        }
+    }
 
     AbstractGeoHexGridTiler(int precision) {
         super(precision);
@@ -53,10 +57,9 @@ abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
         assert bounds.minX() <= bounds.maxX();
         // first check if we are touching just fetch cells
         if (bounds.maxX() - bounds.minX() < 180d) {
-            final long minH3 = H3.geoToH3(bounds.minY(), bounds.minX(), precision);
-            final long maxH3 = H3.geoToH3(bounds.maxY(), bounds.maxX(), precision);
-            if (minH3 == maxH3) {
-                return setValuesFromPointResolution(minH3, values, geoValue);
+            final long singleCell = boundsInSameCell(bounds, precision);
+            if (singleCell > 0) {
+                return setValuesFromPointResolution(singleCell, values, geoValue);
             }
             // TODO: specialize when they are neighbour cells.
         }
@@ -79,7 +82,7 @@ abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
         // Point resolution is done using H3 library which uses spherical geometry. It might happen that in cartesian, the
         // actual point value is in a neighbour cell as well.
         {
-            for (long n : hexRing(h3)) {
+            for (long n : H3.hexRing(h3)) {
                 final GeoRelation relation = relateTile(geoValue, n);
                 valueIndex = maybeAdd(n, relation, values, valueIndex);
                 if (relation == GeoRelation.QUERY_CONTAINS) {
@@ -90,24 +93,23 @@ abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
         return valueIndex;
     }
 
-    private static long[] hexRing(long h3) {
-        long[] neighbours = H3.hexRing(h3);
-        if (h3 == NORTH_POLE) {
-            // Due to equi-rectangular distortion, the hexRing neighbours of the South Pole are insufficient
-            return addNeighbour(neighbours, NORTH_POLE_EXTRA_NEIGHBOUR);
-        } else if (h3 == SOUTH_POLE) {
-            // Due to equi-rectangular distortion, the hexRing neighbours of the South Pole are insufficient
-            return addNeighbour(neighbours, SOUTH_POLE_EXTRA_NEIGHBOUR);
-        } else {
-            return neighbours;
+    private long boundsInSameCell(GeoShapeValues.BoundingBox bounds, int res) {
+        final long minH3 = H3.geoToH3(bounds.minY(), bounds.minX(), res);
+        final long maxH3 = H3.geoToH3(bounds.maxY(), bounds.maxX(), res);
+        if (minH3 != maxH3) {
+            // Normally sufficient to check only bottom-left against top-right
+            return -1;
         }
-    }
-
-    private static long[] addNeighbour(long[] neighbours, long extraNeighbour) {
-        long[] extended = new long[neighbours.length + 1];
-        System.arraycopy(neighbours, 0, extended, 0, neighbours.length);
-        extended[neighbours.length] = extraNeighbour;
-        return extended;
+        if (minH3 == NORTH_POLAR_CELLS[res] || minH3 == SOUTH_POLAR_CELLS[res]) {
+            // But with polar cells we must check the other two corners too
+            final long minMax = H3.geoToH3(bounds.minY(), bounds.maxX(), res);
+            final long maxMin = H3.geoToH3(bounds.maxY(), bounds.minX(), res);
+            if (minMax != minH3 || maxMin != minH3) {
+                return -1;
+            }
+        }
+        // If all checks passed, we can use this cell in an optimization
+        return minH3;
     }
 
     /**
@@ -132,12 +134,11 @@ abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
         // NOTE: When we recurse, we cannot shortcut for CONTAINS relationship because it might fail when visiting noChilds.
         int valueIndex = 0;
         if (bounds.maxX() - bounds.minX() < 180d) {
-            final long minH3 = H3.geoToH3(bounds.minY(), bounds.minX(), 0);
-            final long maxH3 = H3.geoToH3(bounds.maxY(), bounds.maxX(), 0);
-            if (minH3 == maxH3) {
+            final long singleCell = boundsInSameCell(bounds, 0);
+            if (singleCell > 0) {
                 // When the level 0 bounds are within a single cell, we can search that cell and its immediate neighbours
-                valueIndex = setValuesByRecursion(values, geoValue, minH3, 0, valueIndex);
-                for (long n : hexRing(minH3)) {
+                valueIndex = setValuesByRecursion(values, geoValue, singleCell, 0, valueIndex);
+                for (long n : H3.hexRing(singleCell)) {
                     valueIndex = setValuesByRecursion(values, geoValue, n, 0, valueIndex);
                 }
                 return valueIndex;

--- a/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java
+++ b/x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java
@@ -20,10 +20,10 @@ abstract class AbstractGeoHexGridTiler extends GeoGridTiler {
 
     private static final long[] RES0CELLS = H3.getLongRes0Cells();
     // The hexRing neighbours optimization is insufficient at the Poles, so we do a little extra checking
-    private static final long[] NORTH_POLAR_CELLS = new long[H3.MAX_H3_RES];
-    private static final long[] SOUTH_POLAR_CELLS = new long[H3.MAX_H3_RES];
+    private static final long[] NORTH_POLAR_CELLS = new long[H3.MAX_H3_RES + 1];
+    private static final long[] SOUTH_POLAR_CELLS = new long[H3.MAX_H3_RES + 1];
     static {
-        for (int res = 0; res < H3.MAX_H3_RES; res++) {
+        for (int res = 0; res <= H3.MAX_H3_RES; res++) {
             NORTH_POLAR_CELLS[res] = H3.geoToH3(90, 0, res);
             SOUTH_POLAR_CELLS[res] = H3.geoToH3(-90, 0, res);
         }

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexTilerTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexTilerTests.java
@@ -15,6 +15,7 @@ import org.elasticsearch.geo.GeometryTestUtils;
 import org.elasticsearch.geometry.Geometry;
 import org.elasticsearch.geometry.Point;
 import org.elasticsearch.geometry.Rectangle;
+import org.elasticsearch.geometry.utils.StandardValidator;
 import org.elasticsearch.geometry.utils.WellKnownText;
 import org.elasticsearch.h3.H3;
 import org.elasticsearch.xpack.spatial.common.H3CartesianUtil;
@@ -57,16 +58,6 @@ public class GeoHexTilerTests extends GeoGridTilerTestCase {
     @Override
     protected long getCellsForDiffPrecision(int precisionDiff) {
         return UnboundedGeoHexGridTiler.calcMaxAddresses(precisionDiff);
-    }
-
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/92541")
-    public void testGeoGridSetValuesBoundingBoxes_BoundedGeoShapeCellValues() throws Exception {
-        super.testGeoGridSetValuesBoundingBoxes_BoundedGeoShapeCellValues();
-    }
-
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/92562")
-    public void testGeoGridSetValuesBoundingBoxes_UnboundedGeoShapeCellValues() throws Exception {
-        super.testGeoGridSetValuesBoundingBoxes_UnboundedGeoShapeCellValues();
     }
 
     public void testLargeShape() throws Exception {
@@ -126,6 +117,70 @@ public class GeoHexTilerTests extends GeoGridTilerTestCase {
             assertCorner(h3bins, new Point(tile.getMinLon(), tile.getMaxLat()), precision, msg);
             assertCorner(h3bins, new Point(tile.getMaxLon(), tile.getMaxLat()), precision, msg);
         }
+    }
+
+    // Polygons with bounds inside the South Pole cell break a tiler optimization
+    public void testTroublesomeShapeAlmostWithinSouthPole_BoundedGeoShapeCellValues() throws Exception {
+        int precision = 1;
+        String polygon = """
+            POLYGON((180.0 -90.0, 180.0 -73.80002960532788, 1.401298464324817E-45 -73.80002960532788,
+            1.401298464324817E-45 -90.0, 180.0 -90.0))""";
+        GeoBoundingBox geoBoundingBox = new GeoBoundingBox(
+            new GeoPoint(19.585157879020088, 0.9999999403953552),
+            new GeoPoint(-90.0, -26.405694642531472)
+        );
+        Geometry geometry = WellKnownText.fromWKT(StandardValidator.instance(true), true, polygon);
+        GeoShapeValues.GeoShapeValue value = geoShapeValue(geometry);
+        GeoShapeCellValues cellValues = new GeoShapeCellValues(
+            makeGeoShapeValues(value),
+            getBoundedGridTiler(geoBoundingBox, precision),
+            NOOP_BREAKER
+        );
+
+        assertTrue(cellValues.advanceExact(0));
+        int numBuckets = cellValues.docValueCount();
+        int expected = expectedBuckets(value, precision, geoBoundingBox);
+        assertThat("[" + precision + "] bucket count", numBuckets, equalTo(expected));
+    }
+
+    // Polygons with bounds inside the South Pole cell break a tiler optimization
+    public void testTroublesomeShapeAlmostWithinSouthPoleCell_UnboundedGeoShapeCellValues() throws Exception {
+        int precision = 0;
+        String polygon = """
+            POLYGON((1.7481549674935762E-110 -90.0, 180.0 -90.0, 180.0 -75.113250736563,
+            1.7481549674935762E-110 -75.113250736563, 1.7481549674935762E-110 -90.0))""";
+        Geometry geometry = WellKnownText.fromWKT(StandardValidator.instance(true), true, polygon);
+        GeoShapeValues.GeoShapeValue value = geoShapeValue(geometry);
+        GeoShapeCellValues unboundedCellValues = new GeoShapeCellValues(
+            makeGeoShapeValues(value),
+            getUnboundedGridTiler(precision),
+            NOOP_BREAKER
+        );
+
+        assertTrue(unboundedCellValues.advanceExact(0));
+        int numBuckets = unboundedCellValues.docValueCount();
+        int expected = expectedBuckets(value, precision, null);
+        assertThat("[" + precision + "] bucket count", numBuckets, equalTo(expected));
+    }
+
+    // Polygons with bounds inside the North Pole cell break a tiler optimization
+    public void testTroublesomeShapeAlmostWithinNorthPoleCell_UnboundedGeoShapeCellValues() throws Exception {
+        int precision = 1;
+        String polygon = """
+            POLYGON((36.98661841690625 69.44049730644747, 180.0 69.44049730644747,
+            180.0 90.0, 36.98661841690625 90.0, 36.98661841690625 69.44049730644747))""";
+        Geometry geometry = WellKnownText.fromWKT(StandardValidator.instance(true), true, polygon);
+        GeoShapeValues.GeoShapeValue value = geoShapeValue(geometry);
+        GeoShapeCellValues unboundedCellValues = new GeoShapeCellValues(
+            makeGeoShapeValues(value),
+            getUnboundedGridTiler(precision),
+            NOOP_BREAKER
+        );
+
+        assertTrue(unboundedCellValues.advanceExact(0));
+        int numBuckets = unboundedCellValues.docValueCount();
+        int expected = expectedBuckets(value, precision, null);
+        assertThat("[" + precision + "] bucket count", numBuckets, equalTo(expected));
     }
 
     private void assertCorner(long[] h3bins, Point point, int precision, String msg) throws IOException {

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexTilerTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexTilerTests.java
@@ -13,24 +13,17 @@ import org.elasticsearch.common.geo.GeoBoundingBox;
 import org.elasticsearch.common.geo.GeoPoint;
 import org.elasticsearch.geo.GeometryTestUtils;
 import org.elasticsearch.geometry.Geometry;
-import org.elasticsearch.geometry.LinearRing;
 import org.elasticsearch.geometry.Point;
-import org.elasticsearch.geometry.Polygon;
 import org.elasticsearch.geometry.Rectangle;
 import org.elasticsearch.geometry.utils.StandardValidator;
 import org.elasticsearch.geometry.utils.WellKnownText;
-import org.elasticsearch.h3.CellBoundary;
 import org.elasticsearch.h3.H3;
-import org.elasticsearch.h3.LatLng;
 import org.elasticsearch.xpack.spatial.common.H3CartesianUtil;
 import org.elasticsearch.xpack.spatial.index.fielddata.GeoRelation;
 import org.elasticsearch.xpack.spatial.index.fielddata.GeoShapeValues;
 
 import java.io.IOException;
-import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashSet;
 
 import static org.elasticsearch.xpack.spatial.util.GeoTestUtils.geoShapeValue;
 import static org.hamcrest.Matchers.equalTo;

--- a/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexTilerTests.java
+++ b/x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/GeoHexTilerTests.java
@@ -13,17 +13,24 @@ import org.elasticsearch.common.geo.GeoBoundingBox;
 import org.elasticsearch.common.geo.GeoPoint;
 import org.elasticsearch.geo.GeometryTestUtils;
 import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.geometry.LinearRing;
 import org.elasticsearch.geometry.Point;
+import org.elasticsearch.geometry.Polygon;
 import org.elasticsearch.geometry.Rectangle;
 import org.elasticsearch.geometry.utils.StandardValidator;
 import org.elasticsearch.geometry.utils.WellKnownText;
+import org.elasticsearch.h3.CellBoundary;
 import org.elasticsearch.h3.H3;
+import org.elasticsearch.h3.LatLng;
 import org.elasticsearch.xpack.spatial.common.H3CartesianUtil;
 import org.elasticsearch.xpack.spatial.index.fielddata.GeoRelation;
 import org.elasticsearch.xpack.spatial.index.fielddata.GeoShapeValues;
 
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
 
 import static org.elasticsearch.xpack.spatial.util.GeoTestUtils.geoShapeValue;
 import static org.hamcrest.Matchers.equalTo;
@@ -169,6 +176,23 @@ public class GeoHexTilerTests extends GeoGridTilerTestCase {
         String polygon = """
             POLYGON((36.98661841690625 69.44049730644747, 180.0 69.44049730644747,
             180.0 90.0, 36.98661841690625 90.0, 36.98661841690625 69.44049730644747))""";
+        Geometry geometry = WellKnownText.fromWKT(StandardValidator.instance(true), true, polygon);
+        GeoShapeValues.GeoShapeValue value = geoShapeValue(geometry);
+        GeoShapeCellValues unboundedCellValues = new GeoShapeCellValues(
+            makeGeoShapeValues(value),
+            getUnboundedGridTiler(precision),
+            NOOP_BREAKER
+        );
+
+        assertTrue(unboundedCellValues.advanceExact(0));
+        int numBuckets = unboundedCellValues.docValueCount();
+        int expected = expectedBuckets(value, precision, null);
+        assertThat("[" + precision + "] bucket count", numBuckets, equalTo(expected));
+    }
+
+    public void testTroublesomePolarCellLevel1_UnboundedGeoShapeCellValues() throws Exception {
+        int precision = 1;
+        String polygon = "BBOX (-84.24596376729815, 43.36113427778119, 90.0, 83.51476833522361)";
         Geometry geometry = WellKnownText.fromWKT(StandardValidator.instance(true), true, polygon);
         GeoShapeValues.GeoShapeValue value = geoShapeValue(geometry);
         GeoShapeCellValues unboundedCellValues = new GeoShapeCellValues(

